carla/models/catalog/catalog.py

```diff
@@ -0,0 +1,154 @@
+    def __init__(
+        self, data_name, model_type, ext="h5", cache=True, models_home=None, **kws
+    ):
+        self._data_name = data_name
+        self._model = load_model(model_type, data_name, ext, cache, models_home, **kws)
+
+        if ext == "pt":
+            self._backend = "pytorch"
+        elif ext == "h5":
+            self._backend = "tensorflow"
+        else:
+            raise Exception("Model type not in catalog")
+
+        self._name = model_type + "_" + data_name
+
+        if data_name == "adult":
+            self._feature_input_order = [
+                "age",
+                "fnlwgt",
+                "education-num",
+                "capital-gain",
+                "capital-loss",
+                "hours-per-week",
+                "workclass_Private",
+                "marital-status_Non-Married",
+                "occupation_Other",
+                "relationship_Non-Husband",
+                "race_White",
+                "sex_Male",
+                "native-country_US",
+            ]
+        else:
+            raise Exception("Model for dataset not in catalog")
+
+    @property
+    def name(self):
+        """
+        Contains meta information about the model, like name, dataset, structure, etc.
+
+        Returns
+        -------
+        name : String
+            Individual name of the ml model
+        """
+        return self._name
+
+    @property
+    def feature_input_order(self):
+        """
+        Saves the required order of feature as list.
+
+        Prevents confusion about correct order of input features in evaluation
+
+        Returns
+        -------
+        ordered_features : List of String
+            Correct order of input features for ml model
+        """
+        return self._feature_input_order
+
+    @property
+    def backend(self):
+        """
+        Describes the type of backend which is used for the ml model.
+
+        E.g., tensorflow, pytorch, sklearn, ...
+
+        Returns
+        -------
+        backend : String
+            Used framework
+        """
+        return self._backend
+
+    @property
+    def raw_model(self):
+        """
+        Returns the raw ml model built on its framework
+
+        Returns
+        -------
+        ml_model : tensorflow, pytorch, sklearn model type
+            Loaded model
+        """
+        return self._model
+
+    def predict(self, x):
+        """
+        One-dimensional prediction of ml model for an output interval of [0, 1]
+
+        Shape of input dimension has to be always two-dimensional (e.g., (1, m), (n, m))
+
+        Parameters
+        ----------
+        x : np.Array or pd.DataFrame
+            Tabular data of shape N x M (N number of instances, M number of features)
+
+        Returns
+        -------
+        output : np.Array
+            Ml model prediction for interval [0, 1] with shape N x 1
+        """
+
+        assert len(x.shape) == 2
+
+        if self._backend == "pytorch":
```

Comment:
Since we split the behaviour based on the backend implementation, would it make sense to structure the classes instead like this:
```
PytorchModel(MLModel):

TensorflowModel(MLModel):
```
that way the api is clear, e.g. `model.predict()` and `model.predict_proba()` but we could put the backend specific behaviour in their respective implementation of `MLModel`?

Reply:
Sounds great. Later on we will face the problem, that we need to have different ML models for one-hot-encoded data with and without first column dropped, and maybe with different styles of normalization (MinMax, Standard).

Would it make sense to just offer two ML model classes (Tensorflow and Pytorch) in our catalog and keep the inner architecture decisions of encoding and normalization within the class?

Reply:
Yeah true, lets postpone these decision a bit until we have a clearer picture and keep the backend logic in here like it is. Moreover the user would want a uniform interface when loading models I guess and do `my_model = MLModelCatalog("adult", "ann", "pytorch")` instead of `my_model = PytorchModel(...)`